File: cmd/secrt/passphrase.go
package main

import (
	"fmt"
	"io"
	"os"
	"strings"
)

// resolvePassphrase extracts a passphrase from flags using the provided Deps.
// Returns (passphrase, error). Empty passphrase means none requested.
func resolvePassphrase(args parsedArgs, deps Deps) (string, error) {
	count := 0
	if args.passphrasePrompt {
		count++
	}
	if args.passphraseEnv != "" {
		count++
	}
	if args.passphraseFile != "" {
		count++
	}
	if count > 1 {
		return "", fmt.Errorf("specify at most one of --passphrase-prompt, --passphrase-env, --passphrase-file")
	}
	if count == 0 {
		return "", nil
	}

	if args.passphraseEnv != "" {
		p := deps.Getenv(args.passphraseEnv)
		if p == "" {
			return "", fmt.Errorf("environment variable %q is empty or not set", args.passphraseEnv)
		}
		return p, nil
	}

	if args.passphraseFile != "" {
		data, err := os.ReadFile(args.passphraseFile)
		if err != nil {
			return "", fmt.Errorf("read passphrase file: %w", err)
		}
		p := strings.TrimRight(string(data), "\r\n")
		if p == "" {
			return "", fmt.Errorf("passphrase file is empty")
		}
		return p, nil
	}

	// Prompt
	if deps.ReadPass == nil {
		return "", fmt.Errorf("passphrase prompt not available")
	}

	p, err := deps.ReadPass("Passphrase: ", deps.Stderr)
	if err != nil {
		return "", fmt.Errorf("read passphrase: %w", err)
	}
	if p == "" {
		return "", fmt.Errorf("passphrase must not be empty")
	}

	return p, nil
}

// resolvePassphraseForCreate is like resolvePassphrase but prompts for confirmation.
func resolvePassphraseForCreate(args parsedArgs, deps Deps) (string, error) {
	// Check for conflicting flags first (before any I/O)
	count := 0
	if args.passphrasePrompt {
		count++
	}
	if args.passphraseEnv != "" {
		count++
	}
	if args.passphraseFile != "" {
		count++
	}
	if count > 1 {
		return "", fmt.Errorf("specify at most one of --passphrase-prompt, --passphrase-env, --passphrase-file")
	}

	if !args.passphrasePrompt {
		return resolvePassphrase(args, deps)
	}

	if deps.ReadPass == nil {
		return "", fmt.Errorf("passphrase prompt not available")
	}

	p1, err := deps.ReadPass("Passphrase: ", deps.Stderr)
	if err != nil {
		return "", fmt.Errorf("read passphrase: %w", err)
	}
	if p1 == "" {
		return "", fmt.Errorf("passphrase must not be empty")
	}

	p2, err := deps.ReadPass("Confirm passphrase: ", deps.Stderr)
	if err != nil {
		return "", fmt.Errorf("read passphrase confirmation: %w", err)
	}
	if p1 != p2 {
		return "", fmt.Errorf("passphrases do not match")
	}

	return p1, nil
}

func writeError(w io.Writer, jsonMode bool, msg string) {
	if jsonMode {
		fmt.Fprintf(w, `{"error":%q}`+"\n", msg)
	} else {
		fmt.Fprintf(w, "error: %s\n", msg)
	}
}

File: src/passphrase.rs
use std::fs;
use std::io::Write;

use crate::cli::{Deps, ParsedArgs};

/// Extract a passphrase from flags using the provided Deps.
/// Returns (passphrase, error). Empty passphrase means none requested.
pub fn resolve_passphrase(args: &ParsedArgs, deps: &mut Deps) -> Result<String, String> {
    let mut count = 0;
    if args.passphrase_prompt {
        count += 1;
    }
    if !args.passphrase_env.is_empty() {
        count += 1;
    }
    if !args.passphrase_file.is_empty() {
        count += 1;
    }
    if count > 1 {
        return Err(
            "specify at most one of --passphrase-prompt, --passphrase-env, --passphrase-file"
                .into(),
        );
    }
    if count == 0 {
        return Ok(String::new());
    }

    if !args.passphrase_env.is_empty() {
        let p = (deps.getenv)(&args.passphrase_env);
        match p {
            Some(val) if !val.is_empty() => return Ok(val),
            _ => {
                return Err(format!(
                    "environment variable {:?} is empty or not set",
                    args.passphrase_env
                ))
            }
        }
    }

    if !args.passphrase_file.is_empty() {
        let data = fs::read_to_string(&args.passphrase_file)
            .map_err(|e| format!("read passphrase file: {}", e))?;
        let p = data.trim_end_matches(['\r', '\n'].as_ref());
        if p.is_empty() {
            return Err("passphrase file is empty".into());
        }
        return Ok(p.to_string());
    }

    // Prompt
    let p = (deps.read_pass)("Passphrase: ", &mut deps.stderr)
        .map_err(|e| format!("read passphrase: {}", e))?;
    if p.is_empty() {
        return Err("passphrase must not be empty".into());
    }

    Ok(p)
}

/// Like resolve_passphrase but prompts for confirmation on create.
pub fn resolve_passphrase_for_create(args: &ParsedArgs, deps: &mut Deps) -> Result<String, String> {
    // Check for conflicting flags first
    let mut count = 0;
    if args.passphrase_prompt {
        count += 1;
    }
    if !args.passphrase_env.is_empty() {
        count += 1;
    }
    if !args.passphrase_file.is_empty() {
        count += 1;
    }
    if count > 1 {
        return Err(
            "specify at most one of --passphrase-prompt, --passphrase-env, --passphrase-file"
                .into(),
        );
    }

    if !args.passphrase_prompt {
        return resolve_passphrase(args, deps);
    }

    let p1 = (deps.read_pass)("Passphrase: ", &mut deps.stderr)
        .map_err(|e| format!("read passphrase: {}", e))?;
    if p1.is_empty() {
        return Err("passphrase must not be empty".into());
    }

    let p2 = (deps.read_pass)("Confirm passphrase: ", &mut deps.stderr)
        .map_err(|e| format!("read passphrase confirmation: {}", e))?;
    if p1 != p2 {
        return Err("passphrases do not match".into());
    }

    Ok(p1)
}

/// Write an error message to the writer, in JSON or plain format.
pub fn write_error(w: &mut dyn Write, json_mode: bool, msg: &str) {
    if json_mode {
        let _ = writeln!(w, "{{\"error\":{}}}", serde_json::to_string(msg).unwrap());
    } else {
        let _ = writeln!(w, "error: {}", msg);
    }
}
